src/lava/proc/io/bridge/extractor.py

```diff
@@ -0,0 +1,179 @@
+    validate_shape, validate_buffer_size, validate_channel_config, \
+    send_data_blocking, send_data_non_blocking_drop, \
+    recv_empty_blocking, recv_empty_non_blocking_zeros, recv_not_empty_fifo, \
+    recv_not_empty_accumulate
+
+
+class Extractor(AbstractProcess):
+    """Process that gets data out of Lava.
+
+    The Extractor Process exposes a method, recv_data, which enables users to
+    receive data in external Python scripts from the ProcessModel of the
+    Process, while the Lava Runtime is running.
+    Internally, this Process builds a channel (extractor_channel, of type
+    PyPyChannel).
+    The src_port of the channel lives in the ProcessModel.
+    The dst_port of the channel lives in the Process.
+
+    In the ProcessModel, data is received from this Process's InPort,
+    and relayed to the extractor_channel.src_port.
+    When the recv_data is called from the external Python script, data is
+    received from the extractor_channel.dst_port.
+
+    For the sending part of the extractor_channel (src_port), the following
+    property can be parametrized:
+        (1) When the channel buffer is full, sending new data will either:
+            (a) Block until free space is available (until the receiving part
+            receives an item).
+            (b) Not block, and the new data will not be sent.
+
+    For the receiving part of the extractor_channel (recv_port), the following
+    properties can be parametrized:
+        (1) When the channel buffer is empty, receiving will either:
+            (a) Block until an item is available (until the sending part
+            sends an item).
+            (b) Not block, and the received data will be zeros.
+        (2) When the channel buffer is not empty, receiving will either:
+            (a) Receive a single item, being the oldest item put in the
+            channel (FIFO).
+            (b) Receive all items available, accumulated.
+
+    Parameters
+    ----------
+    shape : tuple
+        Shape of the InPort of the Process.
+    buffer_size : optional, int
+        Buffer size of the extractor_channel.
+    extractor_channel_config : optional, ChannelConfig
+        Configuration object specifying how the src_port behaves when the
+        buffer is full and how the dst_port behaves when the buffer is empty
+        and not empty.
+    """
+    def __init__(self,
+                 shape: tuple[int, ...],
+                 buffer_size: ty.Optional[int] = 50,
+                 extractor_channel_config: ty.Optional[ChannelConfig] =
+                 ChannelConfig()) -> None:
+        super().__init__()
+
+        validate_shape(shape)
+        validate_buffer_size(buffer_size)
+        validate_channel_config(extractor_channel_config)
+
+        self._shape = shape
+
+        self._extractor_channel_config = extractor_channel_config
+
+        self._multi_processing = MultiProcessing()
+        self._multi_processing.start()
```

Comment:
This is not ideal. MultiProcessing is meant to be shared by all the Processes which use the same Runtime to ensure that all shared memory is cleaned up consistently.

Instead of doing this, you could defer creation of the PyPyChannel until the runtime is set for this process (`process.run` calls `p.runtime = runtime for p in all_connected_processes`). Once that's done, you can get the runtime._messaging_infrastructure, which will be an instance of MultiProcessing, and use that instead of creating your own instance per Injector/Extractor.

Reply:
I also do not like the fact of using another instance of MultiProcessing here. However, I encountered several hurdles when I tried to make use of the same instance as the one the Runtime uses.

1. If we want to make use of the same MultiProcessing instance as the Runtime, we need to somehow defer the creation of the PyPyChannel to the somewhere between the calls to `_build_message_passing_infrastructure` and `_build_processes` in the `initialize` method of Runtime. This is because the PyPyChannel needs a MultiProcessing instance to be instantiated (and the instance of MultiProcessing is created in `_build_message_passing_infrastructure`) and because the `dst_port` of the created PyPyChannel needs to be passed to the ProcModel through the `proc_params` (and the ProcModel is instantiated with the `proc_params` in `_build_processes`).
   - This is problematic because we have no way of triggering computation to happen in the Process at that time (remember, when the PyPyChannel is created, the `dst_port` needs to be passed to the ProcModel through the `proc_params`, but the Process needs to hold the `src_port` as well!)
   - We could circumvent this by making changes to the Runtime to actually have the possiblity of triggering some callback functions between `_build_message_passing_infrastructure` and `_build_processes` in the `initialize`, but this feels too specific to be added to Runtime.
2. A somehow very similar problem, but yet, arising in a slightly different case, is the following: Let's say you have an Injector->LIF network. And you call run() on the LIF Process, not on the Injector. It's true that internally, a call to run() on the LIF will trigger compilation, then creation of a Runtime, then it will assign the Runtime to all Processes. But in this case, it also needs to trigger some callback in the Injector Process to actually create the PyPyChannel etc...
   - The solution proposed for the first problem could solve this issue... But again, it is very much specific to the Injector and Extractor Processes to be added to the Runtime code.

Reply:
This all makes sense, yes, but there should be ways to work around it, and in my view such a workaround or solution is necessary to avoid the memory issues.

On looking into Runtime._build_channels, it seems like one option would be to use the default channel build mechanism. If you create both an InPort and an OutPort in the proc and connect them together, Runtime will create (and destroy) your PyPyChannel for you, and the ports will be accessible to the proc_model of course because they already get passed from the channel builder (which runs in the user thread) to the subprocess created by the process_builder. So, that only leaves that the process needs to be able to grab its end of the channel, which you should be able to do from Runtime either by searching inside `Runtime._open_ports` (though that might be tricky) or by grabbing the process_builder with `Runtime._get_process_builder_for_process(proc)` then `proc_builder.get_csp_ports()`